src/pages/PopupPage.js

```diff
@@ -178,22 +219,152 @@ function ApproveConnectionForm({ origin, onApprove }) {
+      publicKeys &&
+      publicKeys.some((ownedAccountPubkey) => ownedAccountPubkey.equals(pubkey))
+    ) {
+      return true;
+    }
+    return newAccounts[pubkey.toBase58()] === states.INITIALIZED_AND_OWNED;
+  }
+
+  instructions.forEach((instruction) => {
+    if (!instruction) {
+      unsafe = true;
+    } else if (
+      ['cancelOrder', 'matchOrders', 'newOrder'].includes(instruction.type)
+    ) {
+      // It is always considered safe to cancel orders, match orders, create new orders
+    } else if (instruction.type === 'systemCreate') {
+      let { newAccountPubkey, fromPubkey } = instruction.data;
+      if (!(fromPubkey && newAccountPubkey && fromPubkey.equals(owner))) {
+        unsafe = true;
+      } else {
+        newAccounts[newAccountPubkey.toBase58()] = states.CREATED_AND_OWNED;
+      }
+    } else if (instruction.type === 'initializeAccount') {
+      // New SPL token accounts are only considered safe if they are owned by this wallet and newly created
+      let { ownerPubkey, accountPubkey, mintPubkey } = instruction.data;
+      if (
+        owner &&
+        ownerPubkey &&
+        owner.equals(ownerPubkey) &&
+        accountPubkey &&
+        newAccounts[accountPubkey.toBase58()] === states.CREATED_AND_OWNED
+      ) {
+        newAccounts[accountPubkey.toBase58()] = states.INITIALIZED_AND_OWNED;
+        initializedAccountMints[accountPubkey.toBase58()] = mintPubkey;
+      } else {
+        unsafe = true;
+      }
+    } else if (instruction.type === 'settleFunds') {
+      // Settling funds is only safe if the destinations are owned
+      let { basePubkey, quotePubkey } = instruction.data;
+      if (!isOwned(basePubkey) || !isOwned(quotePubkey)) {
+        unsafe = true;
+      }
+    } else if (instruction.type === 'closeAccount') {
+      // Closing is only safe if closing newly made Wrapped SOL accounts with this wallet as the destination
+      let { sourcePubkey, destinationPubkey, ownerPubkey } = instruction.data;
+      let mintPubkey = initializedAccountMints[sourcePubkey.toBase58()];
+      if (
+        owner &&
+        destinationPubkey &&
+        owner.equals(destinationPubkey) &&
+        mintPubkey &&
+        mintPubkey.equals(WRAPPED_SOL_MINT) &&
+        ownerPubkey &&
+        owner.equals(ownerPubkey) &&
+        sourcePubkey &&
+        newAccounts[sourcePubkey.toBase58()] === states.INITIALIZED_AND_OWNED
+      ) {
+        newAccounts[sourcePubkey.toBase58()] =
+          states.CLOSED_TO_OWNED_DESTINATION;
+      } else {
+        unsafe = true;
+      }
+    } else {
+      unsafe = true;
+    }
+  });
+  return !unsafe;
+}
+
+function ApproveSignatureForm({
+  origin,
+  message,
+  onApprove,
+  onReject,
+  autoApprove,
+}) {
   const classes = useStyles();
   const explorerUrlSuffix = useSolanaExplorerUrlSuffix();
   const connection = useConnection();
   const wallet = useWallet();
+  const [publicKeys] = useWalletPublicKeys();
+  const [safe, setSafe] = useState(false);
```

Comment:
if `safe` is only set by the `useEffect` below it should be a `useMemo` instead of a `useState` plus `useEffect`, to avoid needing to render multiple times